extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -378,32 +406,58 @@ export class Bind {
       });
 
       const onPop = () => this.setState(oldState);
-      return this.setState(result)
-        .then(() => this.getDataForHistory_())
-        .then(data => {
-          this.history_.push(onPop, data);
+      return this.setState(result).then(() => {
+        return this.getDataForHistory_().then(data => {
+          this.historyQueue_.push({op: HistoryOp.PUSH, data, onPop});
+          // Immediately flush queue on "push" to avoid adding latency to
+          // the browser "back" functionality.
+          this.flushHistoryQueue_();
         });
+      });
     });
   }
 
+  /**
+   * @private
+   */
+  flushHistoryQueue_() {
+    for (let i = 0; i < this.historyQueue_.length; i++) {
+      const {op, data, onPop} = this.historyQueue_[i];
+      if (op === HistoryOp.REPLACE) {
+        // Skip consecutive "replace" operations.
+        const next = this.historyQueue_[i + 1];
+        if (next && next.op === HistoryOp.REPLACE) {
+          continue;
```

Comment:
It seems like this could be done better at the push side, which'll prevent this queue from getting massive.

```js
const last = getLast(this.historyQueue_);
if (last && last.op === HistoryOp.REPLACE) {
  last.data = data;
} else {
  this.historyQueue_.push({ op: HistoryOp.REPLACE, data });
}